### PR TITLE
[FIX] website: prevent crash due to `color-mix` in Safari 16.5

### PR DESCRIPTION
**Pull request by Dj0ulo**
__Current behavior before commit:__
The `color` attribute of `hr` is set to `$hr-color`[1]. `$hr-color` is equal to `$border-color`[2] which itself is equal to `--o-border-color`[3].
Finally `--o-border-color` is set using `fade-currentColor()`[4] which uses the `color-mix` css function on `currentColor`[5].

In short, the `color` of `hr` depends on itself. This creates a circular reference that makes Safari 16.5 crash.

This bug is caused by WebKit. There seems to be a commit to fix[6] this bug but it looks like it was not backported to this version of Safari.

__Description of the fix:__
Set `$hr-color` to a fixed value by introducing `--o-hr-color` which is set to the same value of the already existing fallback of `--o-border-color`.

__Steps to reproduce the issue on runbot:__
1. Go to the website shop using Safari 16.5
2. Add an item to the cart
3. Open the cart view
4. Safari crashes

opw-3806725

[1]: https://github.com/odoo/odoo/blob/1154556/addons/web/static/lib/bootstrap/scss/_reboot.scss#L71
[2]: https://github.com/odoo/odoo/blob/1154556/addons/website/static/src/scss/bootstrap_overridden.scss#L333
[3]: https://github.com/odoo/odoo/blob/1154556/addons/website/static/src/scss/bootstrap_overridden.scss#L91
[4]: https://github.com/odoo/odoo/blob/1154556/addons/website/static/src/scss/website.scss#L35
[5]: https://github.com/odoo/odoo/blob/1154556/addons/website/static/src/scss/primary_variables.scss#L2219
[6]: https://github.com/WebKit/WebKit/commit/0d0b884

### CODE REVIEW

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello @Dj0ulo , thanks for your work and good job with the analysis 👍 

Unfortunately the proposed fix it's not really doable 'cause would force the `<hr>` color to `gray-300` regardless by its context → the element will not be visible if rendered into a light-gray container.
Also, by a design standpoint, it make sense that `<hr>` color is based on `currentColor`; it's a correlation we wouldn't want to lose 'cause it achieves the best visual result from the start, without the need of user customizations.

> This bug is caused by WebKit. There seems to be a commit to fix[6](https://github.com/WebKit/WebKit/commit/0d0b884) this bug but it looks like it was not backported to this version of Safari.

This is unfortunate indeed 😕
Since the issue is related to specific versions only, we could try targeting just those ones (?).
The first thing that comes to my mind is to add a [CSS browser hack](http://browserhacks.com/) condition in the `fade-currentColor()` function, but... yeah, it's quite suboptimal, I'm not really sure about that 😕

Maybe @qsm-odoo has a better proposal...

Thanks again!

ps: Are we sure that the bug is affecting `<hr>` only? there are a lot of components using  `$border-color`...

**Comment by Dj0ulo** [comment]
Hi @stefanorigano,

Thanks for your review!

I agree, ideally we should only apply this rule on this particular Safari version. However it looks like there is no browserhacks that can target this one specifically.

> ps: Are we sure that the bug is affecting `<hr>` only? there are a lot of components using `$border-color`...

With my humble research on the dependencies of `$border-color` in the code, I couldn't find any other places that assigns it to a `color` attribute.

Anyway I found [this stackoverflow answer](https://stackoverflow.com/a/6382036/12252679) that explains that we should not use `color` to set the color of an `hr` but rather `border-color` and `background-color`. So a cleaner solution could be to override those instead of relying on [the bootstrap way of doing it](https://github.com/odoo/odoo/blob/1154556a0e681df25e327136a33618b39fd0e090/addons/web/static/lib/bootstrap/scss/_reboot.scss#L71).

Let me know what you think!

**Comment by Dj0ulo** [comment]
> Maybe you could do something with that: https://developer.mozilla.org/fr/docs/Web/CSS/@supports ?

I tried this as a POC
```scss
:root {
    --o-border-color: #{o-color('300')};
}
@supports (color: color-mix(in srgb, currentColor 15%, transparent)) {
    :root {
        --o-border-color: #{fade-currentColor()};
    }
}
```

Safari still crashes. I guess that Safari thinks that it supports this parameters since it is supposed to.

IMO we should stick to [this stackoverflow answer](https://stackoverflow.com/questions/6382023/changing-the-color-of-an-hr-element/6382036#6382036) and not use `color` on `hr`

**Comment by qsm-odoo** [comment]
> IMO we should stick to [this stackoverflow answer](https://stackoverflow.com/questions/6382023/changing-the-color-of-an-hr-element/6382036#6382036) and not use color on hr

If that solves the issue here, then yes probably. Feels strange that it would though :shrug: (and your stackoverflow post does not seem to speak about a problem of color-mix + currentColor :shrug:). Note that the s_hr snippet in website is already doing that: using the border-color instead of color.

File: addons/website/static/src/scss/website.scss [hunk 1]
```diff
@@ -406,6 +406,12 @@ $-header-nav-link-height: $nav-link-height;
     }
 }
 
+hr {
+    border-top: 1px solid $border-color;
```
Comment (stefanorigano) [inline addons/website/static/src/scss/website.scss:410]: Users can override BS variables and, by doing so, customise the website design. 
BS provides an useful set of variables that defines the `<hr>` default design (see [here](https://github.com/twbs/bootstrap/blob/1a6fdfae6be09b09eaced8f0e442ca6f7680a61e/scss/_variables.scss#L663)), we should use them to ensure that overriding BS variables achieves the expected result...

```suggestion
    border-top: $hr-height solid $hr-color;
```

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello @Dj0ulo , thanks for you work.
The proposed approach looks doable to me, one little remark for you in the code.

While investigating your proposal I've discovered another minor issue.
BS uses a solid color + dimm the element by setting `$hr-opacity` to `0.25`.
Our design comes with a color that's "dimmed by default" since:

```scss
$hr-color ← $border-color ← rgba(currentColor, .25)
```
We should set `$hr-opacity: 1 !default` in bootstrap_overridden.scss, could you handle it?
This last issue is not strictly related to your "Safari bug" , but will make your final result shine  😉

Thanks again

**Review by stefanorigano — APPROVED** [review]
Behaves as expected on my side and design is ok.
Thanks @Dj0ulo 👍

**Comment by Dj0ulo** [comment]
@qsm-odoo is it ok for you too? If so, can you r+ this one?

File: addons/website/static/src/scss/website.scss [hunk 2]
```diff
@@ -406,6 +406,12 @@ $-header-nav-link-height: $nav-link-height;
     }
 }
 
+hr {
+    border-top: $hr-height solid $hr-color;
+    color: inherit;
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/website.scss:411]: This deserves a comment. Somethink like
```suggestion
    // Move the set up color as a border color. Indeed, we may use color-mix to
    // define that color as a function of currentColor, in which case Safari is
    // known to crash when using that to set the color property.
    border-top: $hr-height solid $hr-color;
    color: inherit;
```

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@Dj0ulo Sorry for the delay

- What are the `[^x^]` about in your commit message? :thinking: 
- Why is "the bootstrap rule" between `[]` in `[the bootstrap rule][1]`?
- Your `[7]` reference seems wrong (I did not check the others, might be worth double-checking)
- Your commit message does not explain why you changed the hr-opacity, which is actually a change in behavior in stable. I see that SRI asked you too: I guess that's indeed ok as it is minor and probably better most of the time but mention what it is fixing. Really it should be in its own commit in fact.

**Comment by Dj0ulo** [comment]
> * What are the `[^x^]` about in your commit message? 🤔
> * Why is "the bootstrap rule" between `[]` in `[the bootstrap rule][1]`?
> * Your `[7]` reference seems wrong (I did not check the others, might be worth double-checking)

I usually write my commit message in markdown so that the PR message is well formatted. I removed those useless `[]`.

> * Your commit message does not explain why you changed the hr-opacity, which is actually a change in behavior in stable. I see that SRI asked you too: I guess that's indeed ok as it is minor and probably better most of the time but mention what it is fixing. Really it should be in its own commit in fact.

I put this bit in another commit feel free to edit the commit message @qsm-odoo @stefanorigano

File: addons/website/static/src/scss/website.scss (hunk 2 above)
Comment (Dj0ulo) [reply addons/website/static/src/scss/website.scss:414]: Done, but is it really necessary since it's explained in the commit message?

**Comment by Dj0ulo** [comment]
> Well I am entirely for that (I try to too) but in those cases, was that valid markdown? It was real questions: what are those for?

Well in some version of markdown ^x^ puts x in superscript but actually it seems github does not understand it but rather `<sup>x</sup>` (e.g. <sup>x</sup>)
And if you write `[this][1]`, `this` will be shown as a link (e.g. [this][1])

[1]: https://www.youtube.com/watch?v=dQw4w9WgXcQ

File: addons/website/static/src/scss/website.scss (hunk 2 above)
Comment (qsm-odoo) [reply addons/website/static/src/scss/website.scss:414]: Yes, the code should be understandable on its own. Without GitHub history. Having to look the git history to understand the code is a symptom of bad cod or badly documented code. Caring about having a clean git history is important but IMO a 0.1% increase in code quality justify any break in git history (but I know that many don't agree with that at Odoo).

Anyways, in this case: adding the comment has only advantages as it increases code quality without hindering the git history.

**Comment by qsm-odoo** [comment]
> And if you write [this][1], this will be shown as a link (e.g. [this](https://www.youtube.com/watch?v=dQw4w9WgXcQ))

Oh, good to know. Definitely not a markdown expert.

**Review by qsm-odoo — APPROVED** [review]
I push-forced twice
1. Simple rebase
2. Changing both commits (final diff: https://github.com/odoo/odoo/compare/3d593769596adcad50bfe2155d2343119551a0c5..0dec8d242286cd4f077cc89ff3e83a1bc707bfdf)
    - The first commit was still changing the hr-color by mistake. Indeed setting `background-color: transparent` is a change in behavior, that I thus included in the second commit. I had to set up `background-color: $hr-color;` in the first commit to reach the same behavior
    - Adapted s_hr/000.scss accordingly in both commits
    - Fixed whitespaces at the end of line
    - Added some more comments

Thanks for the investigation, let's merge :partying_face: 

@robodoo rebase-ff r+

**Pull request closed without merging**